src/Session.ts

```diff
@@ -0,0 +1,151 @@
+import msgpack5 from 'msgpack5';
+import * as cookie from './Cookie';
+import * as cache from './Cache';
+import { loggerInstance } from './Logger';
+
+const _appDataKey = process.env.SESSION_APP_KEY;
+const _defaultTtl = 60 * 60;
+
+const Session: { [k: string]: any } = {
+
+  cookie: cookie,
+
+  cache: cache,
+
+  /**
+   * Set up default parameters
+   *
+   * @param req - the request object as supplied the the consumer
+   * @return <void>
+   */
+  _setUp: function (req): void {
+    try {
+      this.sessionData = { appData: null, accountData: null };
+      this.sessionId = this.cookie.getSessionId(req.cookies);
+    } catch (err) {
+      loggerInstance().error(err);
+    }
+  },
+
+  /**
+   * Bootstrap a session
+   *
+   * @param req - the request object as supplied the the consumer
+   * @param res - the response object as supplied the the consumer
+   * @return <void>
+   */
+  start: function (req, res): Promise<boolean> {
+    this._setUp(req);
+    return Promise.all(
+      [
+        this.read('appData'),
+        this.read('accountData')
+      ]
+    ).then(([a, b]) => {
+      this.sessionData.appData = JSON.parse(a);
+      this.sessionData.accountData = b ? this.decodeAccountData(b) : b;
+      res.locals.session = this.sessionData;
+      return Promise.resolve(true);
+    }).catch(err => {
+      loggerInstance().error(err);
+      return Promise.reject(false);
+    });
+  },
+
+  /**
+   * Read data from session
+   *
+   * @param type - the type of read to be performed
+   * @return <Promise>
+   */
+  read: function (type: string): Promise<any> {
+    return new Promise((resolve, reject) => {
+      if (type === 'appData') {
+        this.cache.get(_appDataKey)
```

Comment:
Is there a use case to have one data bucket per application?

Reply:
Oh yes.. there's no profit in having app data from multiple applications lumped together in one object especially when we do not need or have use for that session data (from another app). every app is better off managing its own session data, but still have access to the shared account data. I addressed this point in the PR description :)

Reply:
Some apps like FLP appeals frontend needs session data per user so that data is accumulated there until point of submission. Having one Redis key per applications would mean that user A can see session go user B and so on.

Let's talk tomorrow.